File: src/main/scala/com/cloudant/clouseau/AnalyzerService.scala
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.

package com.cloudant.clouseau

import com.yammer.metrics.scala._
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.apache.lucene.analysis.tokenattributes._
import scala.collection.immutable.List
import scalang._
import org.apache.lucene.analysis.Analyzer

class AnalyzerService(ctx: ServiceContext[ConfigurationArgs]) extends Service(ctx) with Instrumented {

  val logger = LoggerFactory.getLogger("clouseau.analyzer")

  override def handleCall(tag: (Pid, Reference), msg: Any): Any = msg match {
    case ('analyze, analyzerConfig: Any, text: String) =>
      SupportedAnalyzers.createAnalyzer(analyzerConfig) match {
        case Some(analyzer) =>
          ('ok, tokenize(text, analyzer))
        case None =>
          ('error, 'no_such_analyzer)
      }
  }

  def tokenize(text: String, analyzer: Analyzer): List[String] = {
    var result: List[String] = List()
    val tokenStream = analyzer.tokenStream("default", text)
    tokenStream.reset()
    while (tokenStream.incrementToken) {
      val term = tokenStream.getAttribute(classOf[CharTermAttribute])
      result = term.toString +: result
    }
    tokenStream.end()
    tokenStream.close()
    result.reverse
  }

}

File: src/main/java/com/cloudant/cloujeau/AnalyzerService.java
package com.cloudant.cloujeau;

import static com.cloudant.cloujeau.OtpUtils.asAtom;
import static com.cloudant.cloujeau.OtpUtils.asBinary;
import static com.cloudant.cloujeau.OtpUtils.asString;
import static com.cloudant.cloujeau.OtpUtils.tuple;

import java.io.IOException;
import java.util.ArrayList;

import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.analysis.TokenStream;
import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;

import com.ericsson.otp.erlang.OtpErlangBinary;
import com.ericsson.otp.erlang.OtpErlangList;
import com.ericsson.otp.erlang.OtpErlangObject;
import com.ericsson.otp.erlang.OtpErlangTuple;

public final class AnalyzerService extends Service {

    private static final OtpErlangBinary[] EMPTY = new OtpErlangBinary[0];

    public AnalyzerService(final ServerState state) {
        super(state, "analyzer");
    }

    @Override
    public OtpErlangObject handleCall(final OtpErlangTuple from, final OtpErlangObject request) throws IOException {
        if (request instanceof OtpErlangTuple) {
            final OtpErlangTuple tuple = (OtpErlangTuple) request;
            if (asAtom("analyze").equals(tuple.elementAt(0))) {
                final OtpErlangObject analyzerConfig = tuple.elementAt(1);
                final OtpErlangBinary text = (OtpErlangBinary) tuple.elementAt(2);
                final Analyzer analyzer = SupportedAnalyzers.createAnalyzer(analyzerConfig);
                if (analyzer != null) {
                    return tuple(asAtom("ok"), tokenize(asString(text), analyzer));
                } else {
                    return tuple(asAtom("error"), asAtom("no_such_analyzer"));
                }
            }
        }
        return null;
    }

    private static OtpErlangList tokenize(final String text, final Analyzer analyzer) throws IOException {
        final ArrayList<OtpErlangBinary> result = new ArrayList<OtpErlangBinary>();
        final TokenStream tokenStream = analyzer.tokenStream("default", text);
        try {
            tokenStream.reset();
            while (tokenStream.incrementToken()) {
                final CharTermAttribute term = tokenStream.getAttribute(CharTermAttribute.class);
                result.add(asBinary(term.toString()));
            }
            tokenStream.end();
        } finally {
            tokenStream.close();
        }
        return new OtpErlangList(result.toArray(EMPTY));
    }

}
